src/mococrw/symmetric_crypto.h

```diff
@@ -0,0 +1,366 @@
+ * #L%
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "openssl_wrap.h"
+
+namespace mococrw
+{
+
+/**
+ * SymmetricCipherMode defines symmetric cipher modes supported by the library.
+ */
+enum class SymmetricCipherMode
+{
+    GCM,
+    CBC,
+    CTR
+};
+
+/**
+ * Supported key lengths for symmetric cipher
+ */
+enum class SymmetricCipherKeySize
+{
+    S_128,
+    S_256
+};
+
+/**
+ * Supported padding types for asymmetric encryption.
+ */
+enum class SymmetricCipherPadding
+{
+    NO,
+    PKCS
+};
+
+/**
+ * Abstract interface for symmetric encryption and decryption.
+ *
+ * This interface defines cipher type and mode agnostic interface for ones-hot and stream-based
+ * operations.
+ *
+ * Interface of an encryptor and decryptor is symmetric and type and mode agnostic. Once created by
+ * a correspondent builder, block or stream-based mode of operation can be used:
+ *
+ * In **one-shot**, input chunk of data is encrypted/decrypted by a single call to update().
+ * Then, the operation is finalized and processed data is returned by finish().
+ *
+ * **Stream-mode** allows to encrypt/decrypt data in chunks. Put a portion of input data into the
+ * cipher using update() and get result data using read(). When all data is processed, finish() must
+ * be called.
+ *
+ * Simple example of encryption:
+ * @code
+ *   // Encryption
+ *   auto encryptor = AESCipherBuilder{SymmetricCipherMode::CTR, SymmetricCipherKeySize::S_256, secretKey}.buildEncryptor();
+ *   encryptor->update(plaintext);
+ *   auto ciphertext = encryptor->finish();
+ *   auto iv = encryptor->getIV();
+ *
+ *   // Decryption
+ *   auto decryptor = AESCipherBuilder{SymmetricCipherMode::CTR, secretKey}.setIV(iv).buildDecryptor();
+ *   decryptor->update(ciphertext);
+ *   auto decryptedText = decryptor->finish();
+ * @endcode
+ */
+class SymmetricCipherI
+{
+public:
+    virtual ~SymmetricCipherI() = default;
+
+    /**
+     * Decrypt or encrypt chunk of data.
+     *
+     * This function encrypts or decrypts chunk of data and places result in the internal buffer.
+     * Processed data can be accessed via read() or/and readAll().
+     *
+     * @param message chunk of data to encrypt/decrypt.
+     */
+    virtual void update(const std::vector<uint8_t> &message) = 0;
+
+    /**
+     * Read portion of encrypted/decrypted data from cipher buffer.
+     *
+     *
+     * Depending on memory strategy and size of read chunk, performance of this method vary. The
+     * current version uses queue of vectors as internal buffer. Therefore, alternate update() and
+     * read() of data blocks of the same size are zero-copy operations.
+     *
+     * @param length size of the requested chunk of data.
+     * @return processed chunk of \c length or smaller.
+     */
+    virtual std::vector<uint8_t> read(size_t length) = 0;
+
+    /**
+     * Read all available data encrypted/decrypted data from the cipher buffer.
```

Comment:
Double "data" in this line, should this be "read all available encrypted/decrypted data"?

Reply:
yes, fixed